### PR TITLE
Setup gradle and create buildSrc

### PR DESCRIPTION
**Pull request by larmie56**
**What does this PR do?**
Setup gradle and create buildSrc

**Description of Task to be completed?**
Setup gradle
Add external lib dependencies
Create the buildSrc folder
Create and configure gradle plugin in the buildSrc folder

### CODE REVIEW

File: buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                sourceCompatibility = JavaVersion.VERSION_1_8
+                targetCompatibility = JavaVersion.VERSION_1_8
+            }
+        }
+
+    }
+
+    private fun importExternalLibs(project: Project) {
+        project.dependencies {
+            add("implementation", Library.coreKtx)
```
Comment (Ezike) [inline buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt]: need to add an extension function on DependencyHandlerScope for `implementation`, `testImplementation`, etc.
Check here:
https://github.com/Ezike/StarWarsSearch-MVI/blob/main/buildSrc/src/main/kotlin/Extensions.kt

File: buildSrc/src/main/kotlin/plugin/LibraryPlugin.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package plugin
+
+import com.android.build.gradle.LibraryExtension
+import org.gradle.api.JavaVersion
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.kotlin.dsl.dependencies
+
+class LibraryPlugin : Plugin<Project> {
```
Comment (Ezike) [inline buildSrc/src/main/kotlin/plugin/LibraryPlugin.kt]: You need to specify if this is an android library or a java library. This looks like an android library. You will need to name it accordingly.

File: buildSrc/src/main/kotlin/BuildType.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+interface BuildType {
```
Comment (Ezike) [inline buildSrc/src/main/kotlin/BuildType.kt:1]: Nice one 👍🏽

File: buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            add("implementation", Library.appCompat)
+            add("implementation", Library.material)
+            add("testImplementation", Library.junit)
+            add("androidTestImplementation", Library.androidxJUnit)
+            add("androidTestImplementation", Library.espresso)
+            // Other external library dependencies
```
Comment (Ezike) [inline buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt]: I don't think you will need these 3rd party libs in the app module, so you can remove them

File: buildSrc/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+kotlin {
+    sourceSets {
+        named("main") {
+            kotlin.apply {
+                srcDir("buildSrc/src/main/kotlin")
+            }
+        }
+    }
+}
```
Comment (Ezike) [inline buildSrc/build.gradle.kts:27]: you don't need this here unless you rename the main source set from java to kotlin

Comment (Ezike) [inline buildSrc/build.gradle.kts:27]: Use the latest version of kotlin, so we can leverage new language features

Comment (Ezike) [inline buildSrc/build.gradle.kts]: remove jcenter repository. it is nearing end of life

File: buildSrc/src/main/kotlin/plugin/LibraryPlugin.kt [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        project.plugins.apply {
+            apply("com.android.library")
+        }
+    }
+
+    private fun applyLibraryExtension(project: Project) {
+        val extension = project.extensions.getByName("android")
+                as? LibraryExtension ?: return
+        extension.apply {
+            compileSdkVersion(AppConfig.compileSdkVersion)
```
Comment (Ezike) [inline buildSrc/src/main/kotlin/plugin/LibraryPlugin.kt]: This is deprecated. Switch to the most recent DSL. Also remove other uses of deprecated APIs

File: buildSrc/src/main/kotlin/plugin/LibraryPlugin.kt [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            buildTypes {
+                named(BuildType.DEBUG) {
+                    isMinifyEnabled = BuildTypeDebug.isMinifyEnabled
+                }
+            }
+
+            compileOptions {
+                sourceCompatibility = JavaVersion.VERSION_1_8
+                targetCompatibility = JavaVersion.VERSION_1_8
+            }
```
Comment (Ezike) [inline buildSrc/src/main/kotlin/plugin/LibraryPlugin.kt]: Use java 11

File: buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+                        getDefaultProguardFile("proguard-android-optimize.txt"),
+                        "proguard-rules.pro"
+                    )
+                }
+            }
+
+            compileOptions {
+                sourceCompatibility = JavaVersion.VERSION_1_8
+                targetCompatibility = JavaVersion.VERSION_1_8
+            }
```
Comment (Ezike) [inline buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt]: Use java 11

Comment (Ezike) [inline buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt]: remove use of deprecated API

Comment (Ezike) [inline buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt]: remove use of deprecated API

File: build.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+// Top-level build file where you can add configuration options common to all sub-projects/modules.
+
+allprojects {
+    repositories {
+        google()
+        mavenCentral()
+        jcenter()
+    }
+}
+
```
Comment (Ezike) [inline build.gradle.kts]: remove jcenter from the repositories as it is reached end of life

File: app/build.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,8 @@
+plugins {
+    id("app")
+}
+
+dependencies {
+    //std lib
```
Comment (Ezike) [inline app/build.gradle.kts]: what's this comment for?

**Review by Ezike — CHANGES_REQUESTED** [review]
Appreciate the work you put in to this PR. Well done.
I have a couple changes and comments in the PR

File: buildSrc/build.gradle.kts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+object Plugin {
+    object Version {
+        const val kotlin: String = "1.4.32"
```
Comment (Ezike) [inline buildSrc/build.gradle.kts]: Use the latest version of kotlin (1.5.x)

File: buildSrc/build.gradle.kts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    }
+}
+
+kotlin {
+    sourceSets {
+        named("main") {
+            kotlin.apply {
+                srcDir("buildSrc/src/main/kotlin")
+            }
+        }
```
Comment (Ezike) [inline buildSrc/build.gradle.kts]: you don't need this here unless you rename the main source set from java to kotlin

**Comment by larmie56** [comment]
> Appreciate the work you put in to this PR. Well done.
> I have a couple changes and comments in the PR

Thank you for the review and your comments, I am currently working on the changes you mentioned in the comments. Thank you once again.

File: buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt (hunk 4 above)
Comment (larmie56) [reply buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt]: Removed some of the 3rd party libs, left a few. I think the project needs the libs I left to build at the moment; so as we create other modules and move them to the modules they are needed, I will remove them from the app module. Is there another (better) way to go about it (I tried to check, I didn't really see anything about it)?

File: buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt [hunk 13]
```diff
@@ -57,21 +62,19 @@ class ApplicationPlugin : Plugin<Project> {
...
-            add("implementation", Library.coreKtx)
-            add("implementation", Library.appCompat)
-            add("implementation", Library.material)
-            add("testImplementation", Library.junit)
-            add("androidTestImplementation", Library.androidxJUnit)
-            add("androidTestImplementation", Library.espresso)
+            implementation(Library.coreKtx)
+            implementation(Library.appCompat)
+            implementation(Library.material)
+
```
Comment (Ezike) [inline buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt]: nice job!

File: buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt [hunk 14]
```diff
@@ -17,9 +22,9 @@ class ApplicationPlugin : Plugin<Project> {
 
     private fun applyPlugins(project: Project) {
         project.plugins.apply {
-            apply("com.android.application")
-            apply("kotlin-android")
-            apply("kotlin-kapt")
+            androidApplication
+            kotlinAndroid
+            kotlinKapt
```
Comment (Ezike) [inline buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt:15]: Nice that you replaced these with the extension functions

File: buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt [hunk 15]
```diff
@@ -57,21 +62,19 @@ class ApplicationPlugin : Plugin<Project> {
             }
 
             compileOptions {
-                sourceCompatibility = JavaVersion.VERSION_1_8
-                targetCompatibility = JavaVersion.VERSION_1_8
+                sourceCompatibility = JavaVersion.VERSION_11
+                targetCompatibility = JavaVersion.VERSION_11
```
Comment (Ezike) [inline buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt]: 👍🏽

File: buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt [hunk 16]
```diff
@@ -35,8 +40,8 @@ class ApplicationPlugin : Plugin<Project> {
                 versionCode(AppConfig.versionCode)
                 versionName(AppConfig.versionName)
 
-                targetSdkVersion(AppConfig.targetSdkVersion)
-                minSdkVersion(AppConfig.minSdkVersion)
+                targetSdk = AppConfig.targetSdkVersion
+                minSdk = AppConfig.minSdkVersion
```
Comment (Ezike) [inline buildSrc/src/main/kotlin/plugin/ApplicationPlugin.kt]: Nice

File: buildSrc/build.gradle.kts [hunk 17]
```diff
@@ -11,24 +11,14 @@ gradlePlugin {
...
-    }
-}
-
-kotlin {
-    sourceSets {
-        named("main") {
-            kotlin.apply {
-                srcDir("buildSrc/src/main/kotlin")
-            }
+            implementationClass = "plugin.AndroidLibraryPlugin"
```
Comment (Ezike) [inline buildSrc/build.gradle.kts]: 👍🏽

File: build.gradle.kts [hunk 18]
```diff
@@ -4,7 +4,6 @@ allprojects {
     repositories {
         google()
         mavenCentral()
-        jcenter()
```
Comment (Ezike) [inline build.gradle.kts]: 👍🏽

**Review by Ezike — APPROVED** [review]
Excellent job with the changes, and use of the extension functions
👏🏽😎

**Comment by larmie56** [comment]
> Excellent job with the changes, and use of the extension functions
> 👏🏽😎

Thank you very much, thanks for the review too; really grateful

**Review by Ezike — COMMENTED** [review]
You can check out my [latest commit](https://github.com/larmie56/ExpenseLogger/commit/1d5a140a56be573f1985bcf8594e68d571b20a8c) that includes some refactor I did.
I deleted the .idea directory cos you don't need to have it on the remote repo, and the IDE will always generate the file anyway.

Let me know if you have any questions

**Comment by larmie56** [comment]
> You can check out my [latest commit](https://github.com/larmie56/ExpenseLogger/commit/1d5a140a56be573f1985bcf8594e68d571b20a8c) that includes some refactor I did.
> I deleted the .idea directory cos you don't need to have it on the remote repo, and the IDE will always generate the file anyway.
> 
> Let me know if you have any questions

Wow! I don't really have any question. I really love looking at your code 💯 . 

What you did with the plugins, especially encapsulating concerns with the BasePlugin and ProjectExtenstions.. wow! I'll still take a look at those two files some more. Thank you very much Tobenna ❤️

**Comment by Ezike** [comment]
> > You can check out my [latest commit](https://github.com/larmie56/ExpenseLogger/commit/1d5a140a56be573f1985bcf8594e68d571b20a8c) that includes some refactor I did.
> > I deleted the .idea directory cos you don't need to have it on the remote repo, and the IDE will always generate the file anyway.
> > Let me know if you have any questions
> 
> Wow! I don't really have any question. I really love looking at your code 💯 .
> 
> What you did with the plugins, especially encapsulating concerns with the BasePlugin and ProjectExtenstions.. wow! I'll still take a look at those two files some more. Thank you very much Tobenna ❤️

you are welcome

**Pull request merged**